platform/platform-impl/src/com/intellij/openapi/progress/impl/ProgressManagerImpl.java

```diff
@@ -73,23 +73,31 @@
 
   // active (i.e. which have executeProcessUnderProgress() method running) indicators which are not inherited from StandardProgressIndicator.
   // for them an extra processing thread (see myCheckCancelledFuture) has to be run to call their non-standard checkCanceled() method
-  private static final Collection<ProgressIndicator> nonStandardIndicators = ConcurrentHashMultiset.create();
+  private static final THashSet<ProgressIndicator> nonStandardIndicators = new THashSet<ProgressIndicator>();
+  private final TObjectProcedure<ProgressIndicator> nonStandardIndicatorCheckLambda = new TObjectProcedure<ProgressIndicator>() {
+    @Override
+    public boolean execute(ProgressIndicator indicator) {
+      try {
+        indicator.checkCanceled();
+      }
+      catch (ProcessCanceledException e) {
+        indicatorCanceled(indicator);
+      }
+      return true;
+    }
+  };
 
   public ProgressManagerImpl() {
     myCheckCancelledFuture = JobScheduler.getScheduler().scheduleWithFixedDelay(new Runnable() {
       @Override
       public void run() {
-        for (ProgressIndicator indicator : nonStandardIndicators) {
-          try {
-            indicator.checkCanceled();
-          }
-          catch (ProcessCanceledException e) {
-            indicatorCanceled(indicator);
-          }
+        THashSet<ProgressIndicator> nonStandardIndicatorsFlushed;
+        synchronized (threadsUnderIndicator) {
+          nonStandardIndicatorsFlushed = nonStandardIndicators;
         }
+        nonStandardIndicatorsFlushed.forEach(nonStandardIndicatorCheckLambda);
       }
```

Comment:
No sence of using local variable of THashSet. We have one object of it. It will never change. **final** modifier(